### ISSUE TITLE
Mark pure pipes dirty when a signal that they read changes 

### ISSUE THREAD

**Issue by jnizet** [angular/angular#56401]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

Currently, if the `transform` method of a pure pipe reads a signal, the pipe won't get reexecuted if the signal changes.

See https://stackblitz.com/edit/stackblitz-starters-eadv1h?file=src%2Fmain.ts as an example.

This is counter-intuitive to me, for the following reasons:

1. The pipe is used in a template, and the template is thus, indirectly, a "client" of the signal, so it seems logical that the template is updated when the signal changes
2. The rules of OnPush have changed to benefit from signals: even if no input of an OnPush component have changed, the component is marked for check when a signal it reads has changed. A pure pipe is very similar, conceptually, to an OnPush component: it's reexecuted only when its arguments change. Since OnPush now also marks a component for check if a signal red by its template has changed, I'd argue that the same should be done for a pure pipe that reads a signal

Making such a change could change the behavior of existing pipes, but the current behavior could be achieved by reading the signal inside an untracked(). And OTOH, pipes that are currently marked as impure because they read a signal could become pure. 

### Proposed solution

Reexecute the pipe even if its argument haven't changed in the case a signal that it reads has changed.

### Alternatives considered

?

**Comment by pkozlowski-opensource** [comment]
While we could make this change, there is a broader question about pure pipes usefulness with signals. Our thinking was that most of the pure pipe usages would be replaced by `computed`. Would this work for your case? If not, could you share more about your particular scenario?

**Comment by pfeileon** [comment]
Well, I guess what you'd really like to do is something like this: `signal() | mappingFunction` and all you have to do is provide a source signal and a pure function. The `computed` part is done for you.

**Comment by e-oz** [comment]
@pkozlowski-opensource 
For me, pipes are more useful than computed, because they allow us to place the business logic and rules in a pipe, and don't copy-pasted as computed() across the codebase.

**Comment by pkozlowski-opensource** [comment]
> allow us to place the business logic and rules in a pipe, and don't copy-pasted as computed() across the codebase.

Same could be said by any regular function invoked from a computed (or any other place)

**Comment by e-oz** [comment]
But `computed()` is declared once and with known params, when a pipe can take params from the template. They are not equal in DX:

```html
@for (product of products; track product.id) {
  if (product | inStock) {
    <div>
       <img [src]="product.images | mostAttractiveImage" alt="."/>
      @if (product | hasFreeShipping ) {
         <label>Free Shipping</label>
       }
       {{ product.price | currency:'USD' }}
    </div>
  }
}
```

It is elegant, simple and effective. How would you replace it all with `computed()`? Maybe there is a way (after 11 hours of working I can't think clearly, sorry), but it will not be as elegant and simple to use.

**Comment by e-oz** [comment]
With `computed()` it would be something like this:

```ts
protected productsWithData = computed(() => {
  return this.products().map((product) => ({
    ...product,
   isInStock: isInStock(product),
  //.. and so on
  }));
});
```

but still, the example with pipes is still easier to modify (and still looks more elegant, especially the currency pipe).

If there is some big benefit of removing pipes - please let me know, maybe I change my opinion. If it's just maintenance cost - then please keep them :)

**Comment by DaSchTour** [comment]
> With `computed()` it would be something like this:
> 
> 
> 
> ```ts
> 
> protected productsWithData = computed(() => {
> 
>   return this.products().map((product) => ({
> 
>     ...product,
> 
>    isInStock: isInStock(product),
> 
>   //.. and so on
> 
>   }));
> 
> });
> 
> ```
> 
> 
> 
> But still, the example with pipes is easier to modify, is not tightly coupled with `products`, and looks more elegant (especially with the currency pipe).
> 
> 
> 
> If there is a significant benefit to removing pipes, please let me know - I might change my opinion. If it's just a maintenance cost, then please keep them :)

Doing many different transformations in one step is something I have seen in many projects and which always lead to issues and poor maintainability.

The worst is, that you may do transformations that aren't even used or you have to put in null or dummy values because for conditional values you actually don't use the data but have to put in values everywhere.

**Comment by e-oz** [comment]
@DaSchTour I almost completely agree :) There is nothing wrong in creating derived reactive values using `computed()` - this approach itself will not lead to issues or poor maintainability. But only while type structures of the derived values are small enough - primitives, or small objects without nesting.

The more complex types we return from `computed()`, the more difficult it becomes to maintain, and the more useless computations we perform. We should create artificial types (interfaces) just to describe the resulting value.

There are other issues in the "`computed()` instead of pipes" approach:
* When your component receives data from multiple different services, it is not always easy to merge different types of data into some derivative value, that your template will consume;
* When you see that your template is too large and you want to move parts of the template into new sub-components, the tight coupling with derived values (when they are large structures) becomes a real issue - you can not just move some part of the template, you also need to re-create derived values for the new sub-components and remove unneeded calculations from the existing derived values, without breaking anything.

**Comment by e-oz** [comment]
@jnizet your example is easy to fix—just return a signal from the pipe:  
https://stackblitz.com/edit/stackblitz-starters-vvjvmh?file=src%2Fmain.ts

But, I 100% agree that the initial example should work.
1. The view is a consumer;
2. The pipe is executed in the reactive context of the view;
3. Therefore, any signals that were read during the pipe execution should become dependencies of the consumer (view).

I'm really surprised this isn't the case. It’s counterintuitive to me too.

**Comment by fabioemoutinho** [comment]
> While we could make this change, there is a broader question about pure pipes usefulness with signals. Our thinking was that most of the pure pipe usages would be replaced by `computed`. Would this work for your case? If not, could you share more about your particular scenario?

@pkozlowski-opensource I feel we need a syntax with a signal/functional approach as an alternative to class based pipes in template.

What we currently have when using signals and pipes together is:
```
{{ mySignal() | multiply: 2 }}
```

But that, compared to how signals simplify things now, feels disconnected from the signals API and verbose (creating a pipe class).

I think I'm looking for a way that allows me to create computed signals in the template based on template parameters and abstracted logic. Example:
```
@Component({
   imports: [multiply], // possible?
   template: `
    @for (item of list; track item) {
      {{ multiply(item.value, 2) }}
      <!-- possibly keep same syntax? -->
      {{ item.value | multiply: 2 }}
    }

    {{ multiply(mySignal(), 2) }}
   `
})
export class MyComponent {
  multiply: SignalPipe<number> = pipe((value: number, multiple: number) => {
     // implicitly uses computed?
     return value * multiple;
  })
}
```

The way I am suggesting it, `pipe` would be a function that returns a signal for the template. Maybe could even be imported as a pipe class is imported currently?

If I just have a function that returns a signal in this way, it would possibly harm the app's performance as Angular would have to reevaluate it on every change detection cycle.
In an Angular implementation perspective, I believe Angular would have to do some sort of magic to avoid additional change detection cycles - just like with pure pipes, where Angular only calls the transform function if inputs have changed.

In other words, `pipe` would basically be an api to create a computed signal for the template, based on its inputs.
Let me know if this makes sense or not. Maybe there are ways to achieve this with `computed` and I am simply not seeing it.

**Comment by Harricpp** [comment]
replace pure pipe usages by `computed`

**Comment by waterplea** [comment]
Don't forget that pipes have access to DI, it's a huge plus when compared to `computed` in DX.

**Comment by pchabros** [comment]
One specific use case can be currency pipe:
```ts
@Pipe({
  name: 'formatCurrency',
  standalone: true,
})
export class FormatCurrencyPipe implements PipeTransform {
  currency = inject(CurrencyService);

  transform(value: number): string {
    return `value ${this.currency.symbol()}`;
  }
}
```
I have currency switch in app, and values won't update on currency change.
I can fix it by wrapping return value with `computed`, but it looks a bit awkward in the template:
```html
<div>{{ (value | formatCurrency)() }}</div>
```

**Comment by pchabros** [comment]
Ok, my example works. I just had to add `pure: false` in `Pipe` configuration.

**Comment by mythridium** [comment]
An example where this could be useful is a custom translation pipe (ignoring that angular has its own translation pipeline for the sake of this example), but this pattern could be applied to similar ideas.

A simplified example, removing some of the unnecessary implementation details.

```ts
@Injectable({ providedIn: 'root' })
export class I18nService {
  private readonly translations = signal({});

  public load(translations: {}) {
    this.translations.set(translations);
  }

  public translate(key: string) {
    return computed(() => this.translations()[key]);
  }
}
```

```ts
@Pipe({ name: 'i18n' })
export class I18nPipe implements PipeTransform {
  private readonly i18n = inject(I18nService);

  public transform(key: string) {
    return this.i18n.translate(key); // returns a signal, the pipe should execute the signal in the html
  }
}
```

```html
{{ 'Example_Key' | i18n }}
```

It would be useful for this pipe to update if either the key changes (current behaviour) or the translations from the service update (for example the user switches to another language) which are sent through a signal.

Of course marking this pipe as impure and reading the signal value in the html would resolve it, but with every single string on the page being translated, and realistically, the trigger points for a change are known (key or translations), it makes little sense to re-evaluate every digest cycle.

If this is already possible with as clean of an implementation as above, please let me know!

**Comment by e-oz** [comment]
@mythridium your pipe example can not be pure, because you expect different results for the same arguments.

**Comment by Racso** [comment]
This is an interesting discussion. My two cents:

I perfectly understand this:

> A pure pipe is very similar, conceptually, to an OnPush component

and I tend to think the same way. However, with the introduction of signals, the "purity" of OnPush changed in a way that may or not be convenient from a formal point of view.

A "pure function" is a [clearly defined concept](https://en.wikipedia.org/wiki/Pure_function) in programming: if you invoke it one million times with the same set of arguments, it MUST return the same value every single time, no matter what. No matter if you change the language of your app, or its theme, or whatever.

Knowing that a function is pure has formal consequences in terms of optimizations, threading, etc.

Pipes that internally react to signals are not pure, by definition, because they will behave differently even under the same set of inputs. This is, _UNLESS we conceptually consider those signals as inputs of the pipe_. In other words: if we consider that any signals the pipe reacts to are actually part of the pipe's inputs list, then a pipe can react to signals and still be considered "pure" (because, under this conception, the signals would be inputs).

That's exactly what happened to OnPush: if we think of signals as being part of the inputs set of an OnPush component, then we can state that OnPush components are "pure" even if they react to signals. But, TBH, I find it a bit far-fetched to consider signals as part of the input set. They're not really "inputs" in the same way that listening to an injected observable isn't an "input".

And that's where I would draw the line: I believe we should NOT think of OnPush as being the same as "pure" anymore. It WAS, but its definition has now diverged. Now, it's just a different concept: OnPush means that the component will be checked when its inputs or signals change. That's it.

And if we draw that conceptual line, then pure pipes not reacting to signals is no longer weird or counterintuitive (because they don't need to behave as OnPush because they are different things). A "pure pipe" would continue working under the strict concept of purity: it only changes if its inputs (**defined as the parameters list of their `transform` function**) change. That, I believe, was the intent of naming pure pipes "pure" in the first place.

**Comment by Racso** [comment]
My other 2 cents: I think some of us (myself included) are focusing too much on the trees ("pure pipes should be pure") and not in the forest ("it would be VERY useful to have pipes that only get re-evaluated when reacting to input and signal changes").

I think the proposal is valid, and @mythridium's example is great. Maybe it's time to have a new type of pipe, circling back partially to the similitudes with components? Pure (only re-evaluates on inputs), OnPush (only re-evaluates on inputs and signals), Default (previously "not-pure"; brute-force mode)?

**Comment by e-oz** [comment]
> are focusing too much on the trees ("pure pipes should be pure")

Pure pipes should be pure. It’s not something unrealistic, it is the bare minimum to keep their behavior predictable.

**Comment by Racso** [comment]
@e-oz Yes, I know. That was the conclusion of my previous comment:

> A "pure pipe" would continue working under the strict concept of purity: it only changes if its inputs (defined as the parameters list of their transform function) change. That, I believe, was the intent of naming pure pipes "pure" in the first place.

What I'm saying is that we are focusing on "pure pipes should be pure" (which, again, is something desirable) and ignoring the _true spirit_ of the request: the usefulness of having [performant] pipes that react to inputs and signals.

In other words: let's ignore the title of this thread ("Mark pure pipes dirty when a signal that they read changes"), and instead imagine that it says "Have some [new] kind of pipes that recalculate only when their inputs or signals change".

We're no longer discussing if pure pipes should have side effects or not (which they SHOULDN'T because they wouldn't be pure anymore), but we're discussing something new: we need [a new type of] pipes that react to signals (but aren't as brute-force as the existing non-pure pipes).

**Comment by e-oz** [comment]
> Have some [new] kind of pipes that recalculate only when their inputs or signals change

Not because I’m a nasty, toxic person, but because that is how pure pipes currently work: you absolutely can do this. Just return a signal. That's it.

**Comment by pdiot** [comment]
You can indeed return a signal, but I feel that a syntax like 
`<span> {{ (myData | pipeReturningSignal)() }}`
is horrendous to look at.

I think a way to cleanly access the value inside the signal from the template would be neat.
Maybe a configuration argument ?

**Comment by waterplea** [comment]
I have a question to the angular team: **do impure pipes that use signals inside count as change detection triggers in the template?** In other words is this:
```html
{{ mySignal() }}
```
same as this:
```html
{{ value | impurePipe }}
```
plus:
```js
transform(value) {
  return this.mySignal()
}
```

So that if this template is buried under multiple levels of OnPush components, does changing `mySignal` value triggers template re-render in both cases?

If the answer is **yes** – this whole issue should be closed as not planned with official answer of **use impure pipes with signals**

**Comment by sod** [comment]
@waterplea Yes they do. Impure pipes that call signals call `markForCheck` if the signal updates. Hence my https://github.com/angular/angular/issues/56401#issuecomment-2724946715.

**Comment by waterplea** [comment]
So I believe there's no reason for this issue to remain open. @JeanMeche do you think something should be updated in the docs to address it?

**Comment by Racso** [comment]
The problem with the "use impure + signals" recommendation is that those impure pipes can ALSO (undesirably) be recomputed in other moments, even if their signals don't change, and even if they're used inside OnPush components.

I'm afraid this thread isn't going anywhere because it focused from the beginning on trying to alter the behavior of pure pipes.

Perhaps we should open a new issue with the general proposal of "providing a way to create pipes that are recalculated if and only if their inputs or signals change, analog to OnPush components"? So the discussion is more focused on the shown uses cases from the start?

**Comment by JeanMeche** [comment]
We discuss this several time with the team, we all agreed that we don't want to change the behavior of pure pipes. They should re-run only if at least of the their arguments change. 

Unpure pipes is what we would recommend today if you'd like to use a signal in the implementation of your Pipe's `transform`.
 
The transform being executed in the reactive context of the template, any signal change will mark the component dirty and ensure CD runs on it (and so does the pipe). 

That being said, this bootstrap a discussion around what should "replace" pipes in a signal world and this topic deserves its own canonical issue.
